### ISSUE TITLE
Example client code

### ISSUE THREAD

**Issue by nrath-js** [tweag/nix-remote-rust#9]
I was super excited to find this crate because I'm looking for a way to do the equivalent of `nix-store -realise <foo>` without having to fork+exec every time.

I am, however, struggling a lot to find a starting point for using it. It would be amazing if there was a trivial example that shows how to connect to a given socket and perform the handshake. (I tried to deduce this from the proxy code, but failed).

**Comment by jneem** [comment]
The way we do it in the proxy is probably the easiest way to get it working, but it does require forking a process: we run `nix-daemon --stdio` and then speak the nix protocol to that process over stdio. This would look something like:

```rust
let mut child = std::process::Command::new("nix-daemon")
    .arg("--stdio")
    .stdin(std::process::Stdio::piped())
    .stdout(std::process::Stdio::piped())
    .spawn()
    .unwrap();

let mut client =
    NixDaemonClient::new(child.stdout.take().unwrap(), child.stdin.take().unwrap()).unwrap();

// Send an op of your choice
let resp_marker = Resp::default();
let op = WorkerOp::OptimizeStore(().into(), resp_marker);
client.send_worker_op_to_daemon(op).unwrap();
client.flush().unwrap();

// Get back the progress/error messages
loop {
    let msg = client.read_error_msg().unwrap();
    if msg == Msg::Last(()) {
        break;
    }
}

// The `resp_marker` just makes sure that the type of the response is what was
// expected from the type of the message. We constructed it as `Resp::default()`,
// but rust should infer it to be a `Resp<u64>` because that's what `WorkerOp::OptimizeStore`
// expects.
let reponse = client.read_build_response_from_daemon(resp_marker).unwrap();
eprintln!("got response: {response:?}");
```

I *think* it might be possible to avoid spawning a process by talking to `/nix/var/nix/daemon-socket/socket` directly, but I haven't tried it. (Of course, to make it work reliably you should check the appropriate nix config and env variables instead of hard-coding the path.)

Taking a step back, the broader issue here is that we're lacking users and use cases, and so our public API has had very little thought. In fact, it was only after trying to write this comment that I realized we were lacking reasonable ways to construct worker ops (see #10). The silver lining here is that I'm eager to improve things based on what you (or others) want to do.

**Comment by nrath-js** [comment]
Thanks for your help! 

Just to confirm: this requires a current Git checkout and doesn't work with the most recent released crate? If so, it'd be great if you could release an updated crate (compiling from Git isn't much of a problem, but it'd be nice to have up-to-date documentation on docs.rs).

I was able to get the example to run and got a "1" response (which I hope indicates success?).

I then modified the example to send an `EnsurePath` command, but got this unexpected error from `read_error_msg`:

```
❯ target/debug/realise /nix/store/5yzw0vhkyszf2d179m0qfkgxmp5wjx4-move-docs.sh
Proxy daemon is: 2.24.8
thread 'main' panicked at src/bin/realise.rs:48:43:
called `Result::unwrap()` on an `Err` value: Deser(Io(Error { kind: UnexpectedEof, message: "failed to fill whole buffer" }))
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Any suggestions?

Here's the full example:
```
use nix_remote::{
    nix_client::NixDaemonClient,
    stderr::Msg,
    worker_op::{Plain, Resp, WorkerOp},
    NixString, StorePath,
};
use std::env;
// use std::io::{Read, Write};
// use std::os::unix::net::UnixStream;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Retrieve the store path from the command-line arguments
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: nix_realise <store-path>");
        std::process::exit(1);
    }

    let store_path = &args[1];
    let store_path = NixString::from(store_path.to_owned());
    let store_path = StorePath(store_path);

    // let socket_path = "/run/nix-daemon.socket";
    // let mut stream = UnixStream::connect(socket_path)?;
    // stream.write_all(&request)?;
    // let mut response = Vec::new();
    // stream.read_to_end(&mut response)?;

    let mut child = std::process::Command::new("nix-daemon")
        .arg("--stdio")
        .stdin(std::process::Stdio::piped())
        .stdout(std::process::Stdio::piped())
        .spawn()
        .unwrap();

    let mut client =
        NixDaemonClient::new(child.stdout.take().unwrap(), child.stdin.take().unwrap()).unwrap();

    // Send an op of your choice
    let resp_marker = Resp::default();
    let op = WorkerOp::EnsurePath(Plain(store_path), resp_marker);
    // let op = WorkerOp::OptimiseStore(().into(), resp_marker);
    client.send_worker_op_to_daemon(&op).unwrap();
    client.flush().unwrap();

    // Get back the progress/error messages
    loop {
        let msg = client.read_error_msg().unwrap();
        if msg == Msg::Last(()) {
            break;
        }
    }

    // The `resp_marker` just makes sure that the type of the response is what was
    // expected from the type of the message. We constructed it as `Resp::default()`,
    // but rust should infer it to be a `Resp<u64>` because that's what `WorkerOp::OptimizeStore`
    // expects.
    let response = client
        .read_build_response_from_daemon(&resp_marker)
        .unwrap();
    eprintln!("got response: {response:?}");

    Ok(())
}
```

**Comment by jneem** [comment]
Hm, I'm having trouble reproducing the crash (using nix-daemon 2.24.10 in case that matters). I've tried calling it with existing store paths, non-existing store paths, and bogus store paths. The bogus store path hangs because it gets a `Msg::Error` instead of a `Msg::Last`, so that probably should get handled separately in the loop.

It would be helpful to see a backtrace, I think, and maybe also to debug-print any responses from `client.read_error_msg` (if there are any messages before the panic).

**Comment by nrath-js** [comment]
Happy to!

```
❯ nix-daemon --version
nix-daemon (Nix) 2.24.8

❯ RUST_BACKTRACE=1 target/debug/realise /nix/store/5yzw0vhkyszf2d179m0qfkgxmp5wjx4-move-docs.sh
Proxy daemon is: 2.24.8
[src/bin/realise.rs:49:9] &msg = Error(
    StderrError {
        typ: [
            69,
            114,
            114,
            111,
            114,
        ],
        level: 0,
        name: [
            69,
            114,
            114,
            111,
            114,
        ],
        message: [
            115,
            116,
            111,
            114,
            101,
            32,
            112,
            97,
            116,
            104,
            32,
            39,
            27,
            91,
            51,
            53,
            59,
            49,
            109,
            53,
            121,
            122,
            119,
            48,
            118,
            104,
            107,
            121,
            115,
            122,
            102,
            50,
            100,
            49,
            55,
            57,
            109,
            48,
            113,
            102,
            107,
            103,
            120,
            109,
            112,
            53,
            119,
            106,
            120,
            52,
            45,
            109,
            111,
            118,
            101,
            45,
            100,
            111,
            99,
            115,
            46,
            115,
            104,
            27,
            91,
            48,
            109,
            39,
            32,
            99,
            111,
            110,
            116,
            97,
            105,
            110,
            115,
            32,
            105,
            108,
            108,
            101,
            103,
            97,
            108,
            32,
            98,
            97,
            115,
            101,
            45,
            51,
            50,
            32,
            99,
            104,
            97,
            114,
            97,
            99,
            116,
            101,
            114,
            32,
            39,
            27,
            91,
            51,
            53,
            59,
            49,
            109,
            45,
            27,
            91,
            48,
            109,
            39,
        ],
        have_pos: 0,
        traces: [],
    },
)
[src/bin/realise.rs:49:9] &msg = Error(
    StderrError {
        typ: [
            69,
            114,
            114,
            111,
            114,
        ],
        level: 0,
        name: [
            69,
            114,
            114,
            111,
            114,
        ],
        message: [
            115,
            116,
            111,
            114,
            101,
            32,
            112,
            97,
            116,
            104,
            32,
            39,
            27,
            91,
            51,
            53,
            59,
            49,
            109,
            53,
            121,
            122,
            119,
            48,
            118,
            104,
            107,
            121,
            115,
            122,
            102,
            50,
            100,
            49,
            55,
            57,
            109,
            48,
            113,
            102,
            107,
            103,
            120,
            109,
            112,
            53,
            119,
            106,
            120,
            52,
            45,
            109,
            111,
            118,
            101,
            45,
            100,
            111,
            99,
            115,
            46,
            115,
            104,
            27,
            91,
            48,
            109,
            39,
            32,
            99,
            111,
            110,
            116,
            97,
            105,
            110,
            115,
            32,
            105,
            108,
            108,
            101,
            103,
            97,
            108,
            32,
            98,
            97,
            115,
            101,
            45,
            51,
            50,
            32,
            99,
            104,
            97,
            114,
            97,
            99,
            116,
            101,
            114,
            32,
            39,
            27,
            91,
            51,
            53,
            59,
            49,
            109,
            45,
            27,
            91,
            48,
            109,
            39,
        ],
        have_pos: 0,
        traces: [],
    },
)
thread 'main' panicked at src/bin/realise.rs:48:43:
called `Result::unwrap()` on an `Err` value: Deser(Io(Error { kind: UnexpectedEof, message: "failed to fill whole buffer" }))
stack backtrace:
   0: rust_begin_unwind
             at /rustc/90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf/library/std/src/panicking.rs:665:5
   1: core::panicking::panic_fmt
             at /rustc/90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf/library/core/src/panicking.rs:74:14
   2: core::result::unwrap_failed
             at /rustc/90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf/library/core/src/result.rs:1700:5
   3: core::result::Result<T,E>::unwrap
             at /usr/local/home/nrath/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs:1104:23
   4: realise::main
             at ./src/bin/realise.rs:48:19
   5: core::ops::function::FnOnce::call_once
             at /usr/local/home/nrath/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

```

**Comment by jneem** [comment]
Ok, so the error message is `store path '\x1b[35;1m5yzw0vhkyszf2d179m0qfkgxmp5wjx4-move-docs.sh\x1b[0m' contains illegal base-32 character '\x1b[35;1m-\x1b[0m'` -- it looks like the hash got truncated, and so nix is trying to interpret a `-` as something in base-32. I've made #11 to make this debug output more readable by default.

This doesn't answer the question about why we're failing to read the error messages. But trying this same truncated store path reproduces the issue on my end, so I should be able to get to the bottom of it. Not until at least tomorrow, though...

**Comment by nrath-js** [comment]
Thanks for investigating! Looks like I dropped a letter in the store path when copying and pasting, grmbl.

I've now gotten it working:
```
❯ RUST_BACKTRACE=1 target/debug/realise /nix/store/5yzw0vhkyszf2d179m0qfkgxmp5wjjx4-move-docs.sh
Proxy daemon is: 2.24.8
[src/bin/realise.rs:49:9] &msg = Last(
    (),
)
got response: 1
```

And I'm happy to report that using a socket to connect to the daemon seems to work as well:
```
    let socket_path = "/run/nix-daemon.socket";
    let stream = UnixStream::connect(socket_path)?;
    let mut client = NixDaemonClient::new(&stream, &stream).unwrap();
```
..gives the same output. So I guess the handshake that's mentioned in the proxy code is not strictly needed? Do you know how likely this is to continue working with future Nix versions?

**Comment by FriedrichAltheide** [comment]
When calling `NixDaemonClient::new` the handshake is already performed: https://github.com/tweag/nix-remote-rust/blob/40e6ca738a47343e4ddf97df566c80c2819dd460/nix-remote/src/nix_client.rs#L23-L32

**Comment by jneem** [comment]
Ok, so the error wasn't caused by an issue with the wire protocol, it was just the nix daemon hanging up after sending those error messages. I've added a `read_error_msg_or_eof` method to make this easier to handle. This method requires the `Read` object to be `BufRead`, because it's hard to test for EOF with just plain `Read`. So with latest main the sample code looks like:

```rust
use nix_remote::{
    nix_client::NixDaemonClient,
    stderr::Msg,
    worker_op::{Plain, Resp, WorkerOp},
    NixString, StorePath,
};
use std::{env, io::BufReader};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Retrieve the store path from the command-line arguments
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: nix_realise <store-path>");
        std::process::exit(1);
    }

    let store_path = &args[1];
    let store_path = NixString::from(store_path.to_owned());
    let store_path = StorePath(store_path);

    let mut child = std::process::Command::new("nix-daemon")
        .arg("--stdio")
        .stdin(std::process::Stdio::piped())
        .stdout(std::process::Stdio::piped())
        .spawn()
        .unwrap();

    let read = BufReader::new(child.stdout.take().unwrap());
    let mut client = NixDaemonClient::new(read, child.stdin.take().unwrap()).unwrap();

    // Send an op of your choice
    let resp_marker = Resp::default();
    let op = WorkerOp::EnsurePath(Plain(store_path), resp_marker);
    client.send_worker_op_to_daemon(&op).unwrap();
    client.flush().unwrap();

    // Get back the progress/error messages
    loop {
        match client.read_error_msg_or_eof().unwrap() {
            Some(Msg::Last(())) => {
                break;
            }
            Some(Msg::Error(e)) => {
                eprintln!("{e:?}");
            }
            Some(_) => {}
            None => {
                eprintln!("hung up");
                return Ok(());
            }
        }
    }

    // The `resp_marker` just makes sure that the type of the response is what was
    // expected from the type of the message. We constructed it as `Resp::default()`,
    // but rust should infer it to be a `Resp<u64>` because that's what `WorkerOp::OptimizeStore`
    // expects.
    let response = client
        .read_build_response_from_daemon(&resp_marker)
        .unwrap();
    eprintln!("got response: {response:?}");

    Ok(())
}

```

**Comment by nrath-js** [comment]
Ah, that makes sense!

So just to confirm, if `client.read_error_msg_or_eof` returns ` Msg::Result`, then this isn't actually a result but more messages should be forthcoming? And only `Msg::Error` and `Msg::Last` means that no further data will come from the daemon?

**Comment by nrath-js** [comment]
Hmm, it also seems that the new `read_error_msg_or_eof` function does not work when using a socket:

```
error[E0599]: the method `read_error_msg_or_eof` exists for struct `NixDaemonClient<&UnixStream, &UnixStream>`, but its trait bounds were not satisfied
  --> src/nix.rs:57:26
   |
57 |         let msg = client.read_error_msg_or_eof()?;
   |                          ^^^^^^^^^^^^^^^^^^^^^
   |
   = note: the following trait bounds were not satisfied:
           `&std::os::unix::net::UnixStream: std::io::BufRead`

```

Is that expected? Should I continue to use `read_error_msg` instead when connecting through a socket?

**Comment by jneem** [comment]
You need to wrap the socket in a `BufReader`. Maybe we could do better in the future, but without the buffer it's tricky to tell the difference between "eof before the message" and "unexpected eof in the middle of the message".

> So just to confirm, if client.read_error_msg_or_eof returns  Msg::Result, then this isn't actually a result but more messages should be forthcoming? And only Msg::Error and Msg::Last means that no further data will come from the daemon?

During successful operation, the stream of messages is terminated by `Msg::Last` (so yes, `Msg::Result` doesn't terminate the stream). For errors, I'm not totally sure what behavior is guaranteed: in this example, we see two identical `Msg::Error`s before eof. In CppNix, the client throws an exception the first time it sees `Msg::Error`, though, so you should probably consider the connection "finished". We should probably have a higher-level API for making this easy, but I'm not sure what it should look like.

> And lastly, do you happen to know if the semantics of the final response value are documented somewhere?

I don't think they are documented anywhere but the CppNix source (which was kind of the main motivation for this project tbh). In this case, `EnsurePath` always [returns 1](https://github.com/NixOS/nix/blob/442a2623e48357ff72c77bb11cf2cf06d94d2f90/src/libstore/daemon.cc#L660). I guess if it isn't successful you're supposed to infer that from the stderr stream.

**Comment by nrath-js** [comment]
This all seems to work nicely, thanks a lot!

**Comment by jneem** [comment]
Great! I think there are some things left to improve still, but I'll close this and open some more specific issues.
